### PR TITLE
fix HttpData unsafe cast

### PR DESCRIPTION
**Pull request by senia-psm**

### CODE REVIEW

**Comment by tusharmath** [comment]
@senia-psm can you please make the PR wrt `main`. We will merge it to the 2.x series internally before releasing.

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala [hunk 1]
```diff
@@ -111,6 +112,11 @@ private[zhttp] trait ServerResponseHandler[R] {
       case HttpData.JavaFile(unsafeGet) =>
         unsafeWriteFileContent(unsafeGet())
         releaseAndRead(jReq)
+
+      case d: HttpData.UnsafeAsync =>
+        rt.unsafeRun(ctx) {
+          writeStreamContent(d.toByteBufStream(ByteBufConfig())).ensuring(ZIO.succeed(releaseAndRead(jReq)))
+        }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]:
```suggestion

      case HttpData.UnsafeAsync(unsafeRun) =>
        unsafeRun {
          _ => c => ctx.writeAndFlush(c.content): Unit
        }
```

I think something simpler like above should suffice for this usecase.

File: zio-http/src/test/scala/zhttp/service/ServerIssue1229Spec.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      } yield assertTrue(resp.status.isSuccess, responseString == "")
+    },
+    test("fromStream") {
+      val stream = ZStream.repeatZIOChunk(ZIO.sleep(10.millis).as(Chunk[String]("A"))).take(10)
+      for {
+        resp           <- callWithContent(HttpData.fromStream(stream))
+        responseString <- resp.bodyAsString
+      } yield assertTrue(resp.status.isSuccess, responseString == "AAAAAAAAAA")
+    },
+  ) @@ timeout(5.seconds)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerIssue1229Spec.scala:60]: We can add a simple test in `ServerSpec.scala` where we are just echoing.

Comment (senia-psm) [reply zio-http/src/test/scala/zhttp/service/ServerIssue1229Spec.scala:60]: I've tried to reproduce the initial issue with `ServerSpec` approach with no luck.

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala (hunk 1 above)
Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: This code does not work for me.

**Comment by senia-psm** [comment]
@tusharmath rebased.

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala (hunk 1 above)
Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: This works, but I believe stream is better:
```scala
      case HttpData.UnsafeAsync(unsafeRun) =>
        unsafeRun { ch => msg =>
          runtime.unsafeRun(ctx)(ChannelFuture.unit(ctx.writeAndFlush(msg.content)))
          if (msg.isLast)
            runtime.unsafeRun(ctx)(ChannelFuture.unit(ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT)))
          else ch.read()
        }
```

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: You are right. There is a missing `ctx.read()` call.
```scala
      case HttpData.UnsafeAsync(unsafeRun) =>
        unsafeRun {
          _ => { c => 
                 ctx.writeAndFlush(c.content)
                 ctx.read(): Unit // This was missing.
               }
        }
```

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: This code fails all my tests (even the one with `HttpData.empty`).

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: This one works, though I'm not sure if it is correct:
```scala
        unsafeRun { ch => msg =>
          ctx.writeAndFlush(msg.content).addListener { (_: Future[_]) =>
            if (msg.isLast) ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT): Unit
            else ch.read()
          }: Unit
        }
```

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: See the new commit

File: zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala [hunk 3]
```diff
@@ -104,9 +105,12 @@ private[zhttp] final class ServerResponseWriter[R](
...
         releaseAndRead(jReq)
 
-      case d: HttpData.UnsafeAsync =>
-        runtime.unsafeRun(ctx) {
-          writeStreamContent(d.toByteBufStream).ensuring(ZIO.succeed(releaseAndRead(jReq)))
+      case HttpData.UnsafeAsync(unsafeRun) =>
+        unsafeRun { ch => msg =>
+          ctx.writeAndFlush(msg.content).addListener { (_: Future[_]) =>
+            if (msg.isLast) ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT): Unit
+            else ch.read()
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala:112]: I think this should also work.
```scala
ctx.writeAndFlush(msg.content)
if(!msg.isLast) ctx.read()
```

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala:112]: Don't we risk to get an OOM this way?
We are reading information from the input channel and pushing it to the output channel without back pressure.
I'll test it with sending 100GB of data without reading the response to find out if we are getting memory leaks this way.

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala:112]: No need for 100GB test - the code above fails all of my tests.

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala:112]: This one passes current tests - I'll check it with 100GB:
```scala
        unsafeRun { ch => msg =>
          ctx.writeAndFlush(msg.content)
          if (msg.isLast) ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT): Unit
          else ch.read()
        }
```

Comment (senia-psm) [reply zio-http/src/main/scala/zhttp/service/ServerResponseWriter.scala:112]: @tusharmath according to my tests Client calls with infinite streams (`HttpData.fromStream`)
 - never returns - there is no way to process response while sending request. It's a blocker for any use-case with infinite request/response streams.
 - leaks memory - server reads request data stream even though nobody is consuming the response data stream. It blocks any proxy-like usage. It seems like there is no back pressure.

I have these issues even with `toByteBufStream` approach.

**Comment by tusharmath** [comment]
@senia-psm continued your work in #1244. 
Closing this PR. Thank you!

**Pull request closed without merging**